src/server/handlers.rs

```diff
@@ -521,6 +521,102 @@ impl<P: Publisher> Handler<P> {
+                        self.send_error(msg.reply, "Internal storage error".to_string())
+                            .await;
+                        return;
+                    }
+                };
+                if stored_manifest.name() != name {
+                    if let Some(deployed_manifest) = stored_manifest.get_version(deployed_version) {
+                        for component in deployed_manifest.spec.components.iter() {
+                            if let Properties::Capability {
+                                properties:
+                                    CapabilityProperties {
+                                        image: image_name, ..
+                                    },
+                            } = &component.properties
+                            {
+                                let image_ref_split: Vec<&str> = image_name.split(':').collect();
+                                if let (Some(&ref_link), Some(&ref_version)) =
+                                    (image_ref_split.first(), image_ref_split.last())
+                                {
+                                    existing_provider_refs
+                                        .insert(ref_link.to_string(), ref_version.to_string());
+                                }
+                            }
+                        }
+                    };
+                }
+            }
+        }
+
+        let staged_model = match req.version.clone() {
+            Some(v) => {
+                if let Some(model) = manifests.get_version(&v) {
+                    model
+                } else {
+                    manifests.get_current()
+                }
+            }
+            None => manifests.get_current(),
+        };
+
+        // Compare if any of the provider refs in the staged model are duplicates
+        for component in staged_model.spec.components.iter() {
+            if let Properties::Capability {
+                properties:
+                    CapabilityProperties {
+                        image: image_name, ..
+                    },
+            } = &component.properties
+            {
+                let image_ref_split: Vec<&str> = image_name.split(':').collect();
+                if let (Some(&ref_link), Some(&ref_version)) =
+                    (image_ref_split.first(), image_ref_split.last())
+                {
+                    if let Some(old_version) = existing_provider_refs.get(ref_link) {
+                        if old_version != ref_version {
+                            error!(
+                                "Provider {image_name} is currently deployed. Forbidden operation.",
+                            );
+                            self.send_error(
+                                msg.reply,
+                                format!(
+                                "Provider {image_name} is currently deployed. Forbidden operation."
```

Comment:
Is there a way here to indicate which model is deployed with the offending provider?

```suggestion
                                "Provider {image_name} is currently deployed with a different version. Forbidden operation.",
                            );
                            self.send_error(
                                msg.reply,
                                format!(
                                "Provider {image_name} is currently deployed with a different version. Forbidden operation."
```